201810/raw-tcp-traffic-shaping-with-istio-1.1.0.md

```diff
@@ -0,0 +1,403 @@
+namespace/istio-system created
+...
+destinationrule.networking.istio.io/istio-policy created
+destinationrule.networking.istio.io/istio-telemetry created 
+```
+
+### 使用Istio代理部署TCP Echo服务
+
+为了演示Istio的路由机制，我们以side-car模式部署`tcp-echo-server` ：
+
+```bash
+$ kubectl apply -f <(istioctl kube-inject -f service.yaml)
+service/tcp-echo-server created
+deployment.extensions/tcp-echo-server-v1 created
+deployment.extensions/tcp-echo-server-v2 created
+```
+
+通过下面的命令来验证服务运行：
+
+```bash
+$ kubectl get pods
+NAME                                  READY     STATUS    RESTARTS   AGE
+tcp-echo-server-v1-78684f5697-sv5r5   2/2       Running   0          56s
+tcp-echo-server-v2-74bf9999c8-hhhf9   2/2       Running   0          56s
+
+$ kubectl logs tcp-echo-server-v1-78684f5697-sv5r5 tcp-echo-server
+listening on [::]:9000, prefix: one
+
+$ kubectl logs tcp-echo-server-v2-74bf9999c8-hhhf9 tcp-echo-server
+listening on [::]:9000, prefix: two
+```
+
+## Istio加权TCP路由
+
+这是本练习的最后一部分，定义`VirtualService`，`DestinationRule`和带有权重路由的 `Gateway`，并验证系统行为。
+
+### 路由配置
+
+创建带有两个`subset`的`DestinationRule`来代表两个版本的 TCP Echo服务。`Gateway` 容许流量通过端口`9000`访问服务。最后，`VirtualService`限定了80%的流量必须被路由到TCP Echo服务的v1版本，20%被路由到v2版本。
+
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: DestinationRule
+metadata:
+  name: destination
+spec:
+  host: tcp-echo-server
+  subsets:
+  - name: v1
+    labels:
+      version: v1
+  - name: v2
+    labels:
+      version: v2
+---
+apiVersion: networking.istio.io/v1alpha3
+kind: Gateway
+metadata:
+  name: gateway
+spec:
+  selector:
+    istio: ingressgateway
+  servers:
+  - port:
+      number: 9000
+      name: tcp
+      protocol: TCP
+    hosts:
+    - "*"
+---
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: route
+spec:
+  hosts:
+  - "*"
+  gateways:
+  - gateway
+  tcp:
+  - match:
+    - port: 9000
+    route:
+    - destination:
+        host: tcp-echo-server
+        port:
+          number: 9000
+        subset: v1
+      weight: 80
+    - destination:
+        host: tcp-echo-server
+        port:
+          number: 9000
+        subset: v2
+      weight: 20
+```
+
+### 部署路由配置
+
+为了让配置生效，复制上面的配置内容并创建文件`route-config.yaml`，用下面的命令进行安装：
+
+```bash
+kubectl apply -f route-config.yaml
+destinationrule.networking.istio.io/destination created
+gateway.networking.istio.io/gateway created
+virtualservice.networking.istio.io/route created
+```
+
+### 验证Istio的TCP路由行为
+
+从本地端口转发请求到`istio-ingressgateway`：
+
+```bash
+$ kubectl get pods -n istio-system | grep ingressgateway
+istio-ingressgateway-7b9bff766d-l478h  1/1  Running  0  30m
+
+$ kubectl port-forward istio-ingressgateway-7b9bff766d-l478h -n istio-system 9000 &
+[1] 69266
+Forwarding from 127.0.0.1:9000 -> 9000
+Forwarding from [::1]:9000 -> 9000
+```
+
+现在可以发送一些请求到加权负载均衡的TCP Echo服务：
+
+```bash
+$ for i in {1..10}; do
+for> docker run -it --rm busybox sh -c '(date; sleep 1) | nc docker.for.mac.localhost 9000'
```

Comment:
Replace `docker.for.mac.localhost 9000` with `192.168.99.100 31400` after identifying the ingress IP. See https://github.com/venilnoronha/venilnoronha.github.io/blob/source/_posts/2018-10-19-raw-tcp-traffic-shaping-with-istio-1.1.0.md#verifying-istios-tcp-routing-behavior.